relay-filter/src/config.rs

```diff
@@ -248,13 +251,117 @@ impl GenericFilterConfig {
+/// the filter's id and the value is the filter itself. The map is converted
+/// back to a list when serializing it, without the filters that were discarded
+/// as duplicates. See examples below.
+///
+/// # Examples
+///
+/// Deserialization:
+///
+/// ```
+/// # use relay_filter::GenericFiltersConfig;
+/// # use insta::assert_debug_snapshot;
+///
+/// let json = r#"{
+///     "version": 1,
+///     "filters": [
+///         {
+///             "id": "filter1",
+///             "isEnabled": false,
+///             "condition": null
+///         },
+///         {
+///             "id": "filter1",
+///             "isEnabled": true,
+///             "condition": {
+///                 "op": "eq",
+///                 "name": "event.exceptions",
+///                 "value": "drop-error"
+///             }
+///         }
+///     ]
+/// }"#;
+/// let deserialized = serde_json::from_str::<GenericFiltersConfig>(json).unwrap();
+/// assert_debug_snapshot!(deserialized, @r#"
+///     GenericFiltersConfig {
+///         version: 1,
+///         filters: {
+///             "filter1": GenericFilterConfig {
+///                 id: "filter1",
+///                 is_enabled: false,
+///                 condition: None,
+///             },
+///         },
+///     }
+/// "#);
+/// ```
+///
+/// Serialization:
+///
+/// ```
+/// # use relay_filter::{GenericFiltersConfig, GenericFilterConfig};
+/// # use relay_protocol::condition::RuleCondition;
+/// # use insta::assert_display_snapshot;
+/// # use indexmap::IndexMap;
+///
+/// let filter = GenericFiltersConfig {
+///     version: 1,
+///     filters: IndexMap::from([(
+///         "filter1".to_owned(),
+///         GenericFilterConfig {
+///             id: "filter1".to_owned(),
+///             is_enabled: true,
+///             condition: Some(RuleCondition::eq("event.exceptions", "drop-error")),
+///         },
+///     )]),
+/// };
+/// let serialized = serde_json::to_string_pretty(&filter).unwrap();
+/// assert_display_snapshot!(serialized, @r#"{
+///   "version": 1,
+///   "filters": [
+///     {
+///       "id": "filter1",
+///       "isEnabled": true,
+///       "condition": {
+///         "op": "eq",
+///         "name": "event.exceptions",
+///         "value": "drop-error"
+///       }
+///     }
+///   ]
+/// }"#);
+/// ```
+#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub struct GenericFiltersConfig {
     /// Version of the filters configuration.
     pub version: u16,
-    /// List of generic filters.
-    pub filters: Vec<GenericFilterConfig>,
+    /// Map of generic filters, sorted by the order in the payload from upstream.
+    ///
+    /// The map contains unique filters, meaning there are no two filters with
+    /// the same id. See struct docs for more details.
+    #[serde(
+        deserialize_with = "deserialize_generic_filters",
+        serialize_with = "serialize_generic_filters"
```

Comment:
nit: See https://serde.rs/field-attrs.html#with

```suggestion
        with = "module_name",
```

Reply:
I'm not sure if this increases readability but applied the change in [45e2f94](https://github.com/getsentry/relay/pull/3161/commits/45e2f94c76d625acc63521cd0f31e3308ffa365f).